### ISSUE TITLE
Refactor Quickstarter tests

### ISSUE THREAD

**Issue by michaelsauter** [opendevstack/ods-quickstarters#414]
**Is your feature request related to a problem?**
The tests are written in Go, one file per quickstarter. That file duplicates a lot of code.

**Describe the solution you'd like**
It's always good to start with duplicating code to see if we really have "the same thing" across all quickstarters. Looking at the codebase now, we see that all quickstarters actually are "the same thing" and that we can make the tests easier to maintain by extracting logic. I would like to see the tests being refactored in such a way that each quickstarter author only needs to provide the "golden files". The logic running the tests can look at the golden files that are present, and then run and verify tests accordingly.

This is not something that needs to be done immediately, but it would be nice to have this eventually to make it easier to maintain the tests, and to add more functionality to them (e.g. cloning the created repository and check the generated files).

**Comment by clemensutschig** [comment]
@michaelsauter here is my thoughts... i Think we should just Pass the golden records down onto the fwrk.. this way you still see in each test whats going on.. and we can run in parallel...

**Comment by michaelsauter** [comment]
Not sure what you mean by "pass the golden records down". Do you mean to have some Go code in each test that reads the files and calls the test framework?

Originally, I had something similar to https://github.com/opendevstack/tailor/pull/215 in mind. Each quickstarter would describe the steps to execute, and the framework would collect the testdata and run its tests against that. Running the tests in parallel is a separate topic I think - it can be achieved by putting the code in separate packages, but also by calling `t.Parallel()` in the test framework.

**Comment by michaelsauter** [comment]
I've been playing a little bit with moving the testing itself to `ods-core`. This has a couple of advantages:

* we remove all duplicated scripts
* we do not need to import ods-core as a Go module in ods-quickstarters
* we remove the code duplication between each quickstarter test (`jenkinsfile_test.go` can be deleted)
* it is super-easy to use this framework against custom quickstarters (not located in `ods-quickstarters`)
* is is easy to see what areas the test actually covers - that was hard to figure out in the Go code (see next section).

So the current experiment has the Go code and the scripts moved to `ods-core/tests/quickstarter`, and can be parameterised with the quickstarter to test. The quickstarter to test just brings the golden files, and a "test instruction" (`steps.yml`), which looks like this currently:
```yml
component: golang-iq-test
steps:
- type: provision
  provisionParams:
    verify:
      jenkinsStages: golden/jenkins-provision-stages.json
- type: build
  buildParams:
    verify:
      jenkinsStages: golden/jenkins-build-stages.json
      sonarScan: golden/sonar-scan.json
      runAttachments:
      - SCRR-{{.ProjectID}}-{{.ComponentID}}.docx
      - SCRR-{{.ProjectID}}-{{.ComponentID}}.md
      testResults: 1
      openShiftResources:
        imageTags:
        - name: "{{.ComponentID}}"
          tag: latest
        buildConfigs:
        - "{{.ComponentID}}"
        imageStreams:
        - "{{.ComponentID}}"
        deploymentConfigs:
        - "{{.ComponentID}}"
        services:
        - "{{.ComponentID}}"
      repoFiles: golden/repo
```

This file is parsed by the test framework, and the steps are executed in order. Some notes:

* All values are Go templates where currently `ProjectID` and `ComponentID` is available as params
* Additional step types can be introduced to e.g. upload files (needed for the release manager)
* Each step allows to override the component / quickstarter, which is needed for the release manager test to work

@clemensutschig @henrjk @gerardcl @renedupont Thoughts?

@clemensutschig Question: Why is the release manager test currently using the prov app API? I can't figure out why. It looks like we could easily use the webhook proxy directly. That way, all quickstarter tests are the same. The prov app API itself seems to be already tested in `ods-core`, so I'd say it's not needed in both places?

**Comment by clemensutschig** [comment]
@michaelsauter  - first of all your proposal is really cool!
regarding your question  - we wanted to have one tests- that tests the component provisioning - via provision app api .. as this is the route that users go .. and its functionality will grow over time .. (E.g. assemble build etc..) ...

**Comment by michaelsauter** [comment]
After our discussion yesterday (@clemensutschig) I came up with the following alternative. We could also keep one Go file per quickstarter, and have it just call a function from ods-core. Advantage would be that we do not need to parse the YML and have a versioned contract with ods-core. Disadvantage is that it is more cumbersome to write, and that it is not so easy to just pass any QS to the framework. Plus, we need to execute the tests from ods-quickstarters then, which necessities some setup scripts which would then be duplicated with ods-core.

Anyhow, I'm just putting it out there (pseudo-code):
```go
func TestBeGolangPlain(t *testing.T) {
	quickstarter.RunTest(quickstarter.TestSteps{
		Component: "golang-iq-test",
		Steps: []quickstarter.TestStep{
			{
				Type: "provision",
				ProvisionParams: quickstarter.ProvisionParams{
					Verify: quickstarter.Verify{
						JenkinsStages: "golden/jenkins-provision-stages.json",
					},
				},
			},
			{
				Type: "build",
				BuildParams: quickstarter.BuildParams{
					Verify: quickstarter.Verify{
						JenkinsStages: "golden/jenkins-build-stages.json",
						SonarScan: "golden/sonar-scan.json",
						RunAttachments: []string{
							"SCRR-{{.ProjectID}}-{{.ComponentID}}.docx",
							"SCRR-{{.ProjectID}}-{{.ComponentID}}.md",
						},
						TestResults: 1,
						OpenShiftResources: quickstarter.OpenShiftResources{
							ImageTags: []quickstarter.ImageTag{
								{
									Name: "{{.ComponentID}}",
									Tag: latest,
								},
							},
							BuildConfigs: []string{"{{.ComponentID}}"},
							ImageStreams: []string{"{{.ComponentID}}"},
							DeploymentConfigs: []string{"{{.ComponentID}}"},
							Services: []string{"{{.ComponentID}}"},
						},
						RepoFiles: "golden/repo",
					},
				},
			},
		},
	})
}
```

**Comment by michaelsauter** [comment]
Closed by https://github.com/opendevstack/ods-core/pull/777 and https://github.com/opendevstack/ods-quickstarters/pull/448.
